renderer,config: add custom DRM modeline support

Akaricchi opened the pull request:
#### Describe your PR, what does it fix/add?

This allows specifying custom display resolutions for the DRM backend. This is useful for display overclocking, working around broken EDIDs, etc. To use this feature, specify a modeline instead of a resolution in the config, for example:

```
monitor = DP-1, modeline 1071.101 3840 3848 3880 3920 2160 2263 2271 2277 +hsync -vsync, 0x0, 1
```

This example is a custom 3840x2160@120Hz mode with tightened timings. I use it because the standard timings don't work with my monitor and GPU combination (M28U with RX580).

The syntax is compatible with Sway and Xorg.


#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

I've been daily driving this for a few weeks and haven't ran into any issues.

#### Is it ready for merging, or does it need work?

I think it is, but probably needs a documentation update as well.

vaxerski commented on src/config/ConfigManager.cpp:
```diff
@@ -469,6 +469,80 @@ void CConfigManager::handleRawExec(const std::string& command, const std::string
+    std::vector<std::string> args;
+    std::stringstream ss;
+
+    for (auto it = modeline.begin();; it++) {
```
useless, you can use `CVarList(modeline, 0, ' ')`

Akaricchi replied:
I thought about it, but the problem with CVarList is that only one character can be a separator. Someone trying to use tabs might get an unexpected surprise (which is also inconsistent with the parser's behavior in other contexts, e.g. even CVarList itself strips both tabs and spaces *around* the separators).

If you are fine with that I can remove the complicated loop, though in my personal opinion it's poor UX.

vaxerski replied:
or you could've just added the functionality to cvarlist. Anyways, I've done that. Rebase against master and use `'s'` as the separator

vaxerski commented on src/render/Renderer.cpp:
```diff
@@ -1515,6 +1515,10 @@ bool CHyprRenderer::applyMonitorRule(CMonitor* pMonitor, SMonitorRule* pMonitorR
         force = true;
     }
 
+    if (pMonitorRule->drmMode.type == DRM_MODE_TYPE_USERDEF) {
```
why? Just store the last modeline

Akaricchi replied:
I didn't want to touch the scary monster if-condition right below. I thought it not important because re-applying the same mode doesn't seem to do anything anyway (that I know of). What do you think of simply removing that whole already-applied check for simplicity?

vaxerski replied:
it's there for a reason, it sets frame skips, monitor reloads, layer reloads, damage and stuff. The "monster if" is not that scary... store the modeline in a string and compare it

vaxerski reviewed (changes requested):
Also please format the file with the .clang-format

Akaricchi commented:
Changed `parseModeLine` to use `CVarList`, implemented same-rule check for custom modelines, and ran clang-format.

Instead of comparing modeline strings, I opted to store `drmModeModeInfo` in `CMonitor` and `memcmp` it with the version from the rule. It's slightly more technically correct, because multiple modeline spellings can parse into identical parameters. It also doesn't require keeping the unparsed modeline in the rule.

Clang-format also caught some stuff not related to my changes, but I filtered them out in the interest of keeping the diff clean.

vaxerski commented on src/config/ConfigManager.cpp:
```diff
@@ -470,6 +470,60 @@ void CConfigManager::handleRawExec(const std::string& command, const std::string
+    auto keyword = args[0];
+    std::transform(keyword.begin(), keyword.end(), keyword.begin(), ::tolower);
+
+    if (keyword != "modeline") {
```
style: no braces around single ifs

vaxerski commented on src/config/ConfigManager.cpp:
```diff
@@ -470,6 +470,60 @@ void CConfigManager::handleRawExec(const std::string& command, const std::string
+    int argno = 1;
+
+    mode.type        = DRM_MODE_TYPE_USERDEF;
+    mode.clock       = stof(args[argno++]) * 1000;
```
`std::`

vaxerski reviewed (changes requested):
minor style changes, rest lgtm

Akaricchi commented:
Done, though I have to say this same file is inconsistent about both of those things.

vaxerski commented:
if it is you can blame my small brane

vaxerski reviewed (approved):
gangsta, thanks!

vaxerski commented:
forgot, could you make a wiki mr? plox

Akaricchi commented:
I'm a little busy currently, but I'll get to that soon-ish.
